File: YukariLauncher/src/main/java/net/kdt/pojavlaunch/modloaders/OFDownloadPageScraper.java
package net.kdt.pojavlaunch.modloaders;

import org.htmlcleaner.HtmlCleaner;
import org.htmlcleaner.HtmlNode;
import org.htmlcleaner.TagNode;
import org.htmlcleaner.TagNodeVisitor;

import java.io.IOException;
import java.net.URL;

/**
 * OptiFineのダウンロードページをスクレイピングして実際のダウンロードURLを取得します。
 */
public class OFDownloadPageScraper implements TagNodeVisitor {
    /**
     * 指定されたURLから実際のダウンロードURLを抽出します。
     * @param urlInput OptiFineのダウンロードページURL
     * @return 実際のダウンロードURL
     * @throws IOException I/Oエラーが発生した場合
     */
    public static String run(String urlInput) throws IOException{
        return new OFDownloadPageScraper().runInner(urlInput);
    }

    private String mDownloadFullUrl;

    /**
     * 内部的なスクレイピング処理を実行します。
     */
    private String runInner(String url) throws IOException {
        HtmlCleaner htmlCleaner = new HtmlCleaner();
        htmlCleaner.clean(new URL(url)).traverse(this);
        return mDownloadFullUrl;
    }

    /**
     * タグノードを訪問し、ダウンロードURLを検出したら抽出します。
     */
    @Override
    public boolean visit(TagNode parentNode, HtmlNode htmlNode) {
        if(isDownloadUrl(parentNode, htmlNode)) {
            TagNode tagNode = (TagNode) htmlNode;
            String href = tagNode.getAttributeByName("href");
            if(!href.startsWith("https://")) href = "https://optifine.net/"+href;
            this.mDownloadFullUrl = href;
            return false;
        }
        return true;
    }

    /**
     * 指定されたノードがダウンロードURLかどうかを判定します。
     */
    public boolean isDownloadUrl(TagNode parentNode, HtmlNode htmlNode) {
        if(!(htmlNode instanceof TagNode)) return false;
        if(parentNode == null) return false;
        TagNode tagNode = (TagNode) htmlNode;
        if(!(parentNode.getName().equals("span")
            && "Download".equals(parentNode.getAttributeByName("id")))) return false;
        return tagNode.getName().equals("a") &&
                "onDownload()".equals(tagNode.getAttributeByName("onclick"));
    }
}

File: YukariLauncher/src/main/java/com/arata/yukarilauncher/feature/mod/modloader/OFDownloadPageScraper.kt
package com.arata.yukarilauncher.feature.mod.modloader

import org.htmlcleaner.HtmlCleaner
import org.htmlcleaner.HtmlNode
import org.htmlcleaner.TagNode
import org.htmlcleaner.TagNodeVisitor
import java.net.URL

/** OptiFineのダウンロードページをスクレイピングして実際のダウンロードURLを取得するクラス。 */
class OFDownloadPageScraper : TagNodeVisitor {
    /** 抽出されたダウンロードURL。 */
    private var mDownloadFullUrl: String? = null

    companion object {
        /** 指定されたURLからダウンロードリンクを抽出して実行する。 @param urlInput OptiFineダウンロードページのURL @return 実際のダウンロードURL */
        fun run(urlInput: String): String {
            return OFDownloadPageScraper().runInner(urlInput)
        }
    }

    /** 内部実行メソッド。HTMLをパースしてダウンロードURLを抽出する。 @param url スクレイピング対象のURL @return 実際のダウンロードURL */
    private fun runInner(url: String): String {
        val htmlCleaner = HtmlCleaner()
        htmlCleaner.clean(URL(url)).traverse(this)
        return mDownloadFullUrl!!
    }

    /** タグノードを訪問し、ダウンロードリンクを検出する。 @param parentNode 親ノード @param htmlNode 現在のHTMLノード @return 走査を続行する場合はtrue */
    override fun visit(parentNode: TagNode?, htmlNode: HtmlNode): Boolean {
        if (isDownloadUrl(parentNode, htmlNode)) {
            val tagNode = htmlNode as TagNode
            var href = tagNode.getAttributeByName("href")
            if (!href.startsWith("https://")) href = "https://optifine.net/$href"
            mDownloadFullUrl = href
            return false
        }
        return true
    }

    /** ノードがダウンロードURLリンクかどうかを判定する。 @param parentNode 親ノード @param htmlNode 判定対象のHTMLノード @return ダウンロードリンクの場合はtrue */
    fun isDownloadUrl(parentNode: TagNode?, htmlNode: HtmlNode): Boolean {
        if (htmlNode !is TagNode) return false
        if (parentNode == null) return false
        val tagNode = htmlNode
        if (!(parentNode.name == "span"
                    && "Download" == parentNode.getAttributeByName("id"))) return false
        return tagNode.name == "a" &&
                "onDownload()" == tagNode.getAttributeByName("onclick")
    }
}
